Added new Plugin Development best practices section (migrated from SDKv2 repo) (#295)
hashicorp/terraform-docs-common: website/docs/plugin/best-practices/provider-code.mdx

---
page_title: Plugin Development - Provider Code
description: Information about writing providers in programming languages other than Go.
---

# HashiCorp Supported Languages and Libraries

## Go

> [Go](https://go.dev/) is currently the only programming language supported for building Terraform providers. The Terraform tooling, documentation, and ecosystem will all assume your provider is being written in and distributed as Go code for the time being.

Reviewer: I'm not sure any of our externally communicated/supported interfaces with providers _assume_ anything about the programming language nor prevent developers from doing something. Most of those interfaces are JSON, which are well supported across languages.

```suggestion
[Go](https://go.dev/) is currently the only programming language supported by HashiCorp for building Terraform providers.
```

### Libraries for writing providers

HashiCorp offers two Go programming language Software Development Kits (SDKs) for building Terraform providers:

#### Terraform Plugin Framework

[Terraform Plugin Framework](/terraform/plugin/framework) is HashiCorp's most recent SDK, and is the current recommended way to develop Terraform providers.

#### Terraform Plugin SDK

[Terraform Plugin SDK](/terraform/plugin/sdkv2), also referred to in documentation as SDKv2, is the prior SDK that many existing Terraform providers use. It is maintained for Terraform versions 1.x and earlier, but we have stopped most feature development so we can focus on improving Terraform Plugin Framework.

### Other provider libraries

#### Terraform Plugin Go

> [Terraform Plugin Go](https://github.com/hashicorp/terraform-plugin-go) provides low-level Go bindings for the [Terraform plugin protocol](/terraform/plugin/terraform-plugin-protocol), for integrations to be built upon. It strives to be a minimal possible abstraction on top of the protocol, only hiding the implementation details of the protocol while leaving its semantics unchanged.

Reviewer: Nit: Typically capitalized

```suggestion
[Terraform Plugin Go](https://github.com/hashicorp/terraform-plugin-go) provides low-level Go bindings for the [Terraform Plugin Protocol](/terraform/plugin/terraform-plugin-protocol), for integrations to be built upon. It strives to be a minimal possible abstraction on top of the protocol, only hiding the implementation details of the protocol while leaving its semantics unchanged.
```

<Highlight>

> `terraform-plugin-go` is a lower-level SDK and is only meant for advanced use cases. Provider developers should consider using the [Terraform Plugin Framework](/terraform/plugin/framework) before using this SDK.

Reviewer: Nit: I think somewhere in the style guide is a recommendation to not end on a callout. I can try to dig up the reference if you need it. To avoid this issue, I think we can just add the "only meant for advanced use cases" part to the paragraph above. The Terraform Plugin Framework mention already says its the recommended way, so we're good there.

Author: Updated ⚡
